### ISSUE TITLE
Better type annotations for `torch.Tensor` subclasses

### ISSUE THREAD

**Issue by matteosantama** [pytorch/pytorch#87371]
### 🚀 The feature, motivation and pitch

In the spirit of #86105, I think `torch.Tensor` type annotations can be improved to better support tensor subclasses.

For example, consider

```python
class Subclass(torch.Tensor):
    pass

subclass = Subclass()

reveal_type(subclass.clone())
# note: Revealed type is "torch._tensor.Tensor"

print(type(subclass.clone())
# __main__.Subclass
```

It's clear that the underlying logic preserves the subclass, but the type annotations don't. There are many methods which would benefit from improved hints: `clone`, `detach`, `new_full`, `new_ones`, `new_empty`, `new_zero`, and `new_tensor`, to name a few.

**Comment by albanD** [comment]
Hey!
Thanks for the suggestion. I think that some of these functions could be updated for sure but in general, the user can return anything they want from these functions when they override them.

That being said I agree that the ones you list should in general return the same type as the input.

We would accept a PR fixing these types!

**Comment by ezyang** [comment]
I'm not sure what this patch would look like. Yes, `new_empty` on a tensor subclass could return the subclass, but it could also not; there's nothing that intrinsically says it works one way or another. 

For methods, I guess I'd be OK with changing our Tensor returns to Self ala https://peps.python.org/pep-0673/ . But this wouldn't always be right, and wouldn't help you for function calls.

**Comment by matteosantama** [comment]
Yes, the patch would look like adding a `Self` annotation to the mentioned `Tensor` _methods_. Or, as it was done in #86105, by annotating the methods with `self: S`. 

> but in general, the user can return anything they want from these functions when they override them.

Should a user override a method to return something else, they will be able to change the annotation in the signature too. But we know what the default (non-overridden) behavior is.

**Comment by jasperjonkhans** [comment]
Hi! Is this typing improvement still wanted? If so, I’d be happy to work on a PR for it.
